text/0000-ssr-rehydration.md

```diff
@@ -0,0 +1,299 @@
+
+- Rather than calling `LWC#createElement`, required that `LWC#hydrateElement` is called to mount the root Lightning Web Component on the client.
+
+- Generate server-side VDOM from the SSR HTML on the client.
+
+- As the initial client-side render progresses, extract state from `data-ssr-state` attributes, instantiate `LightningElements` with extracted state, and generate client-side VDOM.
+
+- Patch server-side VDOM with client-side VDOM.
+  
+  - Where `data-ssr-preserve` is detected, extract the server-side VDOM subtree and store for subsequent client-side renders.
+  
+  - Attach event listeners and non-serializable props.
+  
+  - Resolve unexpected inconsistencies in favor of the client-side VDOM, addressing [Exception C to the happy path](#the-happy-path). With the exception of placeholder components, this case should be rare, since the first CSR will rely on state baked into SSR as `data-ssr-state`.
+
+### `hydrateElement`
+
+In a typical LWC web application, a developer might mount their app using the following pattern:
+
+```javascript
+import { createElement } from 'lwc';
+import MyApp from './my-app';
+
+document
+  .querySelector("#root")
+  .appendChild(createElement('my-app', { is: MyApp }));
+```
+
+After the proposed change, if a developer wants to take advantage of rehydration, a developer would instead do:
+
+```javascript
+import { hydrateElement } from 'lwc';
+import MyApp from './my-app';
+
+hydrateElement(MyApp, document.querySelector('#root')); 
+```
+
+### Server-only Component Support
+
+Addressing [Exception B to the happy path](#the-happy-path) will involve a solution roughly comparable to [React server components](https://reactjs.org/blog/2020/12/21/data-fetching-with-react-server-components.html). While the bulk of this can be implemented in user-land, at a level of abstraction above LWC, the `hydrateElement` patch implementation must support the use-case.
+
+While the initial patch progresses, each SSR VDOM node must be checked for a `data-ssr-preserve` boolean attribute. When identified, the SSR VDOM subtree rooted at that node must be preserved for subsequent client-side renders. Frameworks built on top of LWC can utilize this feature to provide server-only components.
+
+<details open>
+  <summary><b>Example:</b></summary>
+
+An LWC-based framework could provide a new `LightningServerElement` class. Its behavior and underlying implementation could be switched out at build time, as a function of the intended runtime environment.
+
+On the server, a `LightningServerElement` could behave identically to a `LightningElement` with one exception: it would attach a `data-ssr-preserve` attribute to its root node.
+
+For a client build, a Babel plugin could identify any class declaration that extends `LightningServerElement` and replace the entire declaration with `NoopElement`.  That would look something like the following:

```

Comment:
Yes, this would be entirely independent of LWC, really. But documentation/ideas/examples will be important to answer: "How do I make "server components" in LWC like React has?"